### PR TITLE
feat(redis): added shared redis fetch

### PR DESCRIPTION
**Pull request by shahabeshaghi67**
Fixes #???? 

#### PR Checklist

- [ ] Tests
- [ ] Passing Pipeline
- [ ] Documentation

### CODE REVIEW

File: redis/shared.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	res, err = goxy.Decode[K](resStr)
+	if err != nil {
+		return res, err
+	}
+	return res, nil
+}
+
+func fetch[K any](ctx context.Context, client *redis.Client, key string, t time.Duration, f func() (K, error)) {
+	var res K
+	waitKey := fmt.Sprintf("%s:wait", key)
```
Comment (bshramin) [inline redis/shared.go]: We want to use real locks.

File: redis/shared.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+// if it was the case it will not try to fetch
+// but if no one of instances has tried to fetch the result it start to fetch and set a waitKey
+// too inform others
+func SharedFetch[K any](ctx context.Context, client *redis.Client, key string, t time.Duration, f func() (K, error)) (K, error) {
+	var res K
+	redisRes := client.Get(ctx, key)
+	go fetch(ctx, client, key, t, f)
+	err := redisRes.Err()
+	if err != nil {
+		return res, err
```
Comment (bshramin) [inline redis/shared.go:31]: If we get an error, we would probably want to wait for the fetching to complete and then try again.

File: redis/shared.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+		lock = rs.NewMutex(fmt.Sprintf("%s:wait", key))
+		keysList[client][key] = lock
+	}
+	if lock.Until().After(time.Now()) {
+		return
+	}
+	var res K
+	res, err := f()
+	if err != nil {
+		return
```
Comment (bshramin) [inline redis/shared.go:63]: We would want to log the error here.

File: redis/shared.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	}
+	redisSetVal, err := goxy.Encode(res)
+	if err != nil {
+		return
```
Comment (bshramin) [inline redis/shared.go:68]: We would want to log the error here.

File: redis/shared.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+	}
+	err = client.Set(ctx, key, redisSetVal, t).Err()
+	if err == nil {
```
Comment (bshramin) [inline redis/shared.go]:
```suggestion
	if err != nil {
```
If we do as suggested, the readability of the code will increase and we can also log the error.

File: redis/shared.go (hunk 2 above)
Comment (shahabeshaghi67) [reply redis/shared.go:31]: they are two different approaches, if f() takes 1 min for example it's not good.
we can write 2 different functions for them,
one this and another that ensures availibility

**Review by bshramin — APPROVED** [review]

**Pull request merged**